llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp

```diff
@@ -929,11 +929,33 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
 
   IRBuilder<> IRB(InsertBefore);
   Module *M = IRB.GetInsertBlock()->getParent()->getParent();
-  IRB.CreateCall(Intrinsic::getDeclaration(
-                     M, UseShortGranules
-                            ? Intrinsic::hwasan_check_memaccess_shortgranules
-                            : Intrinsic::hwasan_check_memaccess),
-                 {ShadowBase, Ptr, ConstantInt::get(Int32Ty, AccessInfo)});
+
+  // Aarch64 makes it difficult to embed large constants (such as the shadow
+  // offset) in the code. Our intrinsic supports a 16-bit constant (to be left
+  // shifted by 32 bits) - this is not an onerous constraint, since
+  // 1) kShadowBaseAlignment == 32 2) an offset of 4TB (1024 << 32) is
+  // representable, and ought to be good enough for anybody.
+  bool useFixedShadowIntrinsic = true;
+  if (TargetTriple.isAArch64() && ClMappingOffset.getNumOccurrences() > 0 &&
```

Comment:
We better do not touch ClMappingOffset any more.
We copied that into Mapping?

Reply:
Mapping.Offset could be zero because of a manually specified ClMappingOffset option, or because it is Fuchsia or Khwasan. I want to make sure that this optimization is only applied for a manually specified ClMappingOffset option.

```
  if (TargetTriple.isOSFuchsia()) {
    ...
    Offset = 0;
  } else if (ClMappingOffset.getNumOccurrences() > 0) {
    ...  
    Offset = ClMappingOffset;
  } else if (ClEnableKhwasan || InstrumentWithCalls) {
    ...
    Offset = 0;
```

Reply:
Why do we need to exclude them?

Reply:
I'm concerned it may break things, because it will automatically opt Fuchsia and Khwasan into using the fixedshadow intrinsic (unlike on Aarch64 HWASan, where this patch only takes effect if they opt-in to fixed shadows). I don't have the infrastructure set up to test my change on Fuchsia or Khwasan.

Reply:
This is not a reason to complicate with unnecessary branches. Unless you have a good reason, please make it as generic as possible.

Reply:
A new reason: to avoid needing the shadow base in the intrinsic, I've added a getFixedShadowBase() function, but that is in the llvm namespace and doesn't have access to Mapping.